cpp/src/arrow/filesystem/filesystem_utils.h

```diff
@@ -0,0 +1,90 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_FILESYSTEM_FACTORY_H
+#define ARROW_FILESYSTEM_FACTORY_H
+
+#include <memory>
+#include <string>
+#include "arrow/filesystem/filesystem.h"
+
+namespace arrow {
+
+namespace fs {
+
+/// \class FileSystemOptions
+/// \brief Options to create FileSystem instance
+struct ARROW_EXPORT FileSystemOptions {
+ public:
+  /// \brief Create default Filesystem options
+  ///
+  /// \param[out] options Made FileSystemOptions
+  /// \return Status
+  static Status Make(std::shared_ptr<FileSystemOptions>* options);
+
+  /// \brief Configure parameters for HDFS
+  ///
+  /// \param[in] use_hdfs3 a switch for using hdfs or hdfs3 lib
+  /// \param[in] replication replication number of hdfs
+  void ConfigureHDFSOptions(bool use_hdfs3, int replication = 1);
```

Comment:
actually, instead of duplicating `use_hdfs3` in both FileSystemOptions and HdfsOptions, it would probably be better FileSystemOptions could simply contain an instance of HdfsOptions (required iff fs_type is inferred as hdfs)